### ISSUE TITLE
Can model fields be optional ("schemaless" models)?

### ISSUE THREAD

**Issue by omanizer** [vuex-orm/vuex-orm#559]
First off, this project is fantastic.  I discovered it a few hours ago and I've already started migrating a project over to it.  I've created a lot of manual CRUD implementations on top of Vuex and they have been OK, but this makes everything so much easier.  I plan to use this a lot in the future.  Thank you!

I've noticed that when I don't declare *every* field, the fields I don't declare don't show up in the record.

This is a request from a lazy person (I'm asking for a friend :) )... but can model fields be optional?  I  trust and am satisfied with what the API returns to me.  Up until now all of my Vuex stores have just been arrays of objects, no need for a definition, so all properties of returned data from the API end up in the store.  The main benefit I see from having to declare fields is casting and default values, which are nice, but I'd also love the ability to just declare a model and skip having to declare fields entirely.

Maybe something like `static schemaless = true` or something.  I'd love to be able to Model.insert something and have it just add every property to the store.

This may also help with polymorphic objects where fields aren't defined (granted, I'd almost always put schemaless data inside of an object property like "data" or something anyway since it would be a JSON(B) column in the DB, but, for example, mongo would allow schemaless at the root level).

Thanks for your consideration.

**Comment by cuebit** [comment]
Glad to hear you're reaping the rewards, @omanizer.

Vuex ORM relies on data schema(s) in order to normalize and process data and relationships. Without said schema(s) this would be impossible.

However, I like the idea of collecting non-declared fields that are detached from the entity but recorded nonetheless. For example, and correct me if I'm diverging from your statement, declared fields can be considered "safe attributes" – granting all the functionalities Vuex ORM has to offer, and non-declared fields could be collected as "unsafe attributes", which can be stored in parallel and can be retrieved by `$model.getUnsafeAttributes()` or similar (otherwise returning an object as key:value).

While this can be achieved by extending models with supporting modules, I guess the trade off would be the cost of processing additional payload when inserting and retrieving from the store.

**Comment by omanizer** [comment]
I think what you're saying is in the right direction and would work fine as along as these "unsafe attributes" could be adjacent to the declared fields.  Here's how I would see it being used in my case:

I define a "User" model with these fields:
```
static fields () {
  return {
    id: this.attr(),
    name: this.attr(),
    email: this.attr()
  }
}
```
When I make a request from the server I would get something like this:
```
GET /v1/users/1
{
  "data": {
    "id": 1,
    "name": "Omanizer"
    "email": "me@example.com",
    "created_at": "2010-01-01T00:00:00Z"
  }
}
```

At that point I would have a user record.  It would be really nice to be able to do any of these:
```
user.id // This works and would return "1"
user.created_at // Currently doesn't work
user.getUnsafeAttributes().created_at // This is good but I like the above better.
```
Relations and casting are both nice, but it would be an acceptable trade-off in my case to be able to assume that I would not get either of those functionalities unless I actually declared the field.  Generally I'd use this on fields that I'm not using in relations anyway (I'll never join on the created_at field).

**Comment by cuebit** [comment]
@omanizer sorry for the delayed response. Been thinking about this issue in-depth. 

Wondering if you've managed to solve this issue yourself?

The general consensus is that the desired feature is way off the purpose of Vuex ORM and something that you won't expect in common ORM patterns. Granted that the "unsafe attributes" is something that is supported in places, hence the mention, but isn't a feature that will be brought into scope any time soon. It is, however, something that could be worth tackling via a plugin if there is enough demand for it.

**Comment by cuebit** [comment]
Closing issue due to inactivity. Please open a new issue if there's anything further to add. 👍

**Comment by omanizer** [comment]
Sorry, your message on 3/21 flew under the radar.  I agree that most ORMs don't directly support this, especially in context of a SQL database that doesn't support it, where tables always have defined columns.  Even Postgres/MySQL json columns reside under a defined column (which Vuex-ORM works fine for).  For a "schemaless" db such as Mongo, this isn't the case.  And while the "Mongoose" library isn't considered an actual "ORM", it is used to that same end quite a bit and thus making this library compatible with it could be useful.

Mongoose does support fields that aren't directly defined when the Schema is initialized (and it's very important that it does).  While you can't directly access them (directly typing `user.created_at` in the above example), it is possible to do both of the following for a field that isn't pre-defined:
```
user.set('created_at', '2000-01-01T00:00:00Z');
user.get('created_at')
```

In one project I'm adopting this library for (a CRM with the ability to specify custom data fields for users), the endpoint (call call it /api/contacts) on the server returns custom root-level fields that differ (significantly) between different accounts using the service (Some people want to store a pet's name, car color, or other custom data).  It's worth saying that out of about 200 endpoints, this is the ONLY one that doesn't have a definite root-level schema but this endpoint makes up about 90% of the data in this service, so it's very significant.  It's arguable that a better design pattern would have been to do server-side what I'm going to outline below (if only I could go back 5-6 years, haha).

To get this to work just define the Vuex-ORM model with these fields: 
```
{
  _id: this.attr(null),
  data: this.attr(null) // this will be an object
}
```

Then when the server returns the data that looks something like this:
```
{
  _id: 'deadbeef',
  first_name: 'Joe',
  last_name: 'Cool',
  favorite_color: 'Red'
}
```
...I'm just intercepting the request / response in the axios plugin for that particular endpoint and moving everything except for "_id" into the "data" object.  Then, when I call any of the HTTP save methods I end up moving all the data to root level props before sending to server.  This seems to work decent but I do worry that there will come more issues in the future from it.  I don't love that it places a lot of this logic into the HTTP transport rather than the model itself.  But, it works and it'll do OK for the time being.

I also can define instance methods on the model class for `get` and `set`, so it's pretty much the same functionality as mongoose but I won't be binding any v-models to these records under the data path, and it's actually better in my case to not do that.

Anyway, thanks for your consideration!
